app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts

```diff
@@ -0,0 +1,222 @@
+   * @param args.state - Initial state to set on this controller.
+   * @param args.getMetaMetricsId - A function that returns the current MetaMetrics ID.
+   */
+  constructor({
+    dataDeletionService,
+    messenger,
+    state,
+    getMetaMetricsId,
+  }: {
+    dataDeletionService: PublicInterface<DataDeletionService>;
+    messenger: MetaMetricsDataDeletionControllerMessenger;
+    state?: Partial<MetaMetricsDataDeletionState>;
+    getMetaMetricsId: () => string | null;
+  }) {
+    // Call the constructor of BaseControllerV2
+    super({
+      messenger,
+      metadata,
+      name: controllerName,
+      state: { ...defaultState, ...state },
+    });
+    this.#getMetaMetricsId = getMetaMetricsId;
+    this.#dataDeletionService = dataDeletionService;
+    this.#registerMessageHandlers();
+  }
+
+  /**
+   * Constructor helper for registering this controller's messaging system
+   * actions.
+   */
+  #registerMessageHandlers(): void {
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:createMetaMetricsDataDeletionTask`,
+      this.createMetaMetricsDataDeletionTask.bind(this),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:updateDataDeletionTaskStatus`,
+      this.updateDataDeletionTaskStatus.bind(this),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:setHasMetaMetricsDataRecorded`,
+      this.setHasMetaMetricsDataRecorded.bind(this),
+    );
+  }
+
+  /**
+   * Creating the delete regulation using source regulation
+   *
+   */
+  async createMetaMetricsDataDeletionTask(): Promise<void> {
+    const metaMetricsId = this.#getMetaMetricsId();
+    if (!metaMetricsId) {
+      throw new Error('MetaMetrics ID not found');
+    }
+
+    const { data } =
+      await this.#dataDeletionService.createDataDeletionRegulationTask(
+        metaMetricsId,
+      );
+    this.update((state) => {
+      state.metaMetricsDataDeletionId = data?.regulateId;
+      state.metaMetricsDataDeletionDate = Date.now();
+      return state;
+    });
+    await this.updateDataDeletionTaskStatus();
+  }
+
+  /**
+   * To check the status of the current delete regulation.
+   */
+  async updateDataDeletionTaskStatus(): Promise<void> {
+    const deleteRegulationId = this.state.metaMetricsDataDeletionId;
+    if (deleteRegulationId.length === 0) {
+      throw new Error('Delete Regulation id not found');
+    }
+
+    const {
+      data: { regulation },
+    } = await this.#dataDeletionService.fetchDeletionRegulationStatus(
+      deleteRegulationId,
+    );
+
+    this.update((state) => {
+      state.metaMetricsDataDeletionStatus = regulation.overallStatus;
+      return state;
+    });
+  }
+
+  /**
+   * To records whether data was collected after last deletion
+   *
+   * @param record - boolean value to determine data is being recorded.
+   */
+  setHasMetaMetricsDataRecorded(record: boolean): void {
```

Comment:
What is this method for? Generally speaking, "setter" methods on controllers are something we should be avoiding, they are an anti-pattern. Controllers should be responsible for guaranteeing the correctness/consistency of their own state, and setters make that impossible (the caller can use them to set state incorrectly).

Reply:
I see that in #24571, this is called when the MetaMetrics toggle is enabled. Even more concerningly, we're relying on the UI to tell us when this happens.

Rather than asking the caller to know exactly when to call this, the deletion controller should listen for state changes of the toggle itself. Then it can react to the toggle being flipped immediately, without any caller needing to know about it.